### PR TITLE
Support for getBuyIntentExtraParams()

### PR DESCRIPTION
**Pull request by ratm**
This PR addresses issue #241, adding support for getBuyIntentExtraParams() as this is the recommended method by Google.
https://developer.android.com/google/play/billing/billing_reference.html#getBuyIntent

> The method is a variant of the getBuyIntent() method, and takes an additional extraParams parameter.

More info here: https://developer.android.com/google/play/billing/billing_reference.html#getBuyIntentExtraParams

### CODE REVIEW

**Comment by serggl** [comment]
instead of adding a new parameter to existing function,lets just add another function

**Comment by ratm** [comment]
@serggl I'm not sure to understand your remark. 
Could you elaborate? Thanks!

**Comment by ratm** [comment]
@serggl Please let me know if my changes properly addressed your feedback. Thanks!

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 1]
```diff
@@ -461,21 +522,31 @@ private boolean purchase(Activity activity, List<String> oldProductIds, String p
...
 			if (oldProductIds != null && purchaseType.equals(Constants.PRODUCT_TYPE_SUBSCRIPTION))
 			{
-				bundle =
-						billingService.getBuyIntentToReplaceSkus(Constants.GOOGLE_API_SUBSCRIPTION_CHANGE_VERSION,
-																 contextPackageName,
-																 oldProductIds,
-																 productId,
-																 purchaseType,
-																 purchasePayload);
+				int apiVersion = Constants.GOOGLE_API_EXTRA_PARAMS_SUPPORTED_VERSION;
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: Im worried that this API might not be supported, but same time v5 API (`GOOGLE_API_SUBSCRIPTION_CHANGE_VERSION`) may be. So its kind of a breaking change. We have `isSubscriptionUpdateSupported()` method that was checking if v5 API is supported, but we dont have one to check the availability of this newly introduced API.
So I guess there should be something in this method that checks for API version and uses backward compatible method if needed

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 2]
```diff
@@ -461,21 +522,31 @@ private boolean purchase(Activity activity, List<String> oldProductIds, String p
...
+						purchaseType, purchasePayload, extraParamsBundle);
 			}
 			else
 			{
-				bundle = billingService.getBuyIntent(Constants.GOOGLE_API_VERSION,
-													 contextPackageName,
-													 productId,
-													 purchaseType,
-													 purchasePayload);
+				bundle = billingService.getBuyIntentExtraParams(Constants.GOOGLE_API_EXTRA_PARAMS_SUPPORTED_VERSION,
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: same here. if we change API version from v3 to v6 we must ensure we have a  backward compatibility plan

**Comment by ratm** [comment]
@serggl I'm now using isBillingSupportedExtraParams() to use the new API only if available.

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 3]
```diff
@@ -461,21 +589,48 @@ private boolean purchase(Activity activity, List<String> oldProductIds, String p
...
+				}
 			}
 			else
 			{
-				bundle = billingService.getBuyIntent(Constants.GOOGLE_API_VERSION,
-													 contextPackageName,
-													 productId,
-													 purchaseType,
-													 purchasePayload);
+				if( extraParamsBundle == null) // API v3
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: can we remove a leading whitespace here?

File: library/src/main/java/com/anjlab/android/iab/v3/Constants.java [hunk 4]
```diff
@@ -19,6 +19,8 @@
 {
 	public static final int GOOGLE_API_VERSION = 3;
 	public static final int GOOGLE_API_SUBSCRIPTION_CHANGE_VERSION = 5;
+	public static final int GOOGLE_API_EXTRA_PARAMS_SUPPORTED_VERSION = 6;
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/Constants.java]: seem like this constant is not used for right now

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 5]
```diff
@@ -364,6 +413,50 @@ public boolean isSubscriptionUpdateSupported()
 		return isSubsUpdateSupported;
 	}
 
+	public boolean isSubscriptionWithExtraParamsSupported(Bundle extraParams)
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: a little Javadoc here would be 👍

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 6]
```diff
@@ -433,6 +526,35 @@ public boolean updateSubscription(Activity activity, List<String> oldProductIds,
...
+	 * @param productId     the new subscription id
+	 * @param developerPayload the developer payload
+	 * @param extraParams  A bundle object containing extra parameters to pass to getBuyIntentExtraParams()
+	 * @see <a href="https://developer.android.com/google/play/billing/billing_reference.html#getBuyIntentExtraParams">extra
+	 * params documentation on developer.android.com</a>
+	 * @return {@code false} if {@code oldProductIds} is not {@code null} AND change subscription
+	 * is not supported.
+	 */
+	public boolean updateSubscription(Activity activity, List<String> oldProductIds,
+													 String productId, String developerPayload, Bundle extraParams)
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: argument list here seem to be off from the indentation

File: library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java [hunk 7]
```diff
@@ -461,21 +589,48 @@ private boolean purchase(Activity activity, List<String> oldProductIds, String p
...
 			if (oldProductIds != null && purchaseType.equals(Constants.PRODUCT_TYPE_SUBSCRIPTION))
 			{
-				bundle =
-						billingService.getBuyIntentToReplaceSkus(Constants.GOOGLE_API_SUBSCRIPTION_CHANGE_VERSION,
-																 contextPackageName,
-																 oldProductIds,
-																 productId,
-																 purchaseType,
-																 purchasePayload);
+				if(extraParamsBundle == null) // API v3
```
Comment (serggl) [inline library/src/main/java/com/anjlab/android/iab/v3/BillingProcessor.java]: Api v5 actually :)

**Comment by serggl** [comment]
This now looks good!
last thing I could think of is that the changes made should be somehow reflected in the README

**Comment by serggl** [comment]
Thank you :+1: Will release this early next week

**Comment by ratm** [comment]
Sounds great! Thank you @serggl!

**Pull request merged**